File: app/src/main/java/com/huhx0015/androidbooster/architecture/mvp/BaseContract.java
package com.huhx0015.androidbooster.architecture.mvp;

/**
 * Reference: https://www.codeproject.com/Articles/1098822/Learn-Android-MVP-Pattern-By-Example
 */

public interface BaseContract {

    /** VIEW INTERFACE _________________________________________________________________________ **/

    // View: Defines the methods that the concrete View aka Fragment will implement. This way you can
    // proceed to create and test the Presenter without worrying about Android-specific components
    // such as Context.
    interface View extends BaseView<Presenter> {}

    /** PRESENTER INTERFACE ____________________________________________________________________ **/

    // Presenter: Defines the methods that the concrete Presenter class will implement. Also known
    // as user actions, this is where the business logic for the app is defined.
    interface Presenter extends BasePresenter {}

    /** REPOSITORY INTERFACE ___________________________________________________________________ **/

    // Repository: Defines the methods that the concrete persistence class will implement. This way
    // the Presenter does not need to be concerned about how data is persisted.
    interface Repository {}
}

File: app/src/main/java/com/huhx0015/androidbooster/architecture/viper/BaseContract.kt
package com.huhx0015.androidbooster.architecture.viper

/**
 * Reference: https://cheesecakelabs.com/blog/using-viper-architecture-android/
 */

interface BaseContract {

    /** VIEW INTERFACE _________________________________________________________________________  */

    interface View

    /** INTERACTOR INTERFACE ___________________________________________________________________  */

    interface Interactor

    interface InteractorOutput

    /** PRESENTER INTERFACE ____________________________________________________________________  */

    interface Presenter

    /** ENTITY INTERFACE _______________________________________________________________________  */

    interface Entity

    /** ROUTER INTERFACE _______________________________________________________________________  */

    interface Router
}